config/v1/types_infrastructure.go

```diff
@@ -1277,6 +1307,45 @@ type VSpherePlatformTopology struct {
 	Template string `json:"template,omitempty"`
 }
 
+// VSphereFailureDomainAffinity contains the vCenter cluster vm-host group (virtual machine and host types)
+// and the vm-host affinity rule that together creates a affinity configuration for vm-host based zonal.
+// This configuration within vCenter creates the required association between a failure domain, virtual machines
+// and ESXi hosts to create a vm-host based zone.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.hostGroup) : !has(self.hostGroup)",message="hostGroup is required when type is HostGroup, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'HostGroup' ?  has(self.vmGroup) : !has(self.vmGroup)",message="vmGroup is required when type is HostGroup, and forbidden otherwise"
```

Comment:
Is this right? Normally there are different type values that map 1:1 with fields in the spec

If all of these fields are only required when type is hostGroup, they should all be in a hostGroup struct

If they are then optional for other values of type, you should make optional members that match up with those and include the fields for their type, with the correct validations